pyfar/dsp/dsp.py

```diff
@@ -2669,3 +2669,173 @@ def normalize(signal, reference_method='max', domain='auto',
+        >>>
+        >>> ax.set_xlabel('time lag in samples')
+        >>> ax.set_ylabel('auto correlation')
+        >>> ax.legend()
+
+    Compute the lags (delay) that are required to maximize the
+    cross-correlation between a one and multi-dimensional signal
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>> import numpy as np
+        >>>
+        >>> # one-dimensional signal: impulse with zero delay
+        >>> signal_1 = pf.signals.impulse(5, 0)
+        >>>
+        >>> # multi-dimension signal with cshape = (2, 2):
+        >>> # impulses with non-zero delays
+        >>> delays = np.array([[0, 1], [2, 3]], dtype=int)
+        >>> signal_2 = pf.signals.impulse(5, delays)
+        >>>
+        >>> _, _, argmax = pf.dsp.correlate(signal_1, signal_2, 'linear')
+
+    In this case the lags correspond to the negative delays:
+    ``argmax = [[0, -1], [-2, -3]]``.
+    """
+
+    # check input
+    if type(signal_1) is not pyfar.Signal or \
+            type(signal_2) is not pyfar.Signal:
+        raise TypeError("signal_1 and signal_2 must be pyfar.Signal objects")
+
+    if signal_1.complex != signal_2.complex:
+        raise ValueError(("Both signals must be complex or real valued. "
+                          "Mixtures are not allowed."))
+
+    if mode not in ['linear', 'cyclic']:
+        raise ValueError(f"mode is '{mode}' but must be 'linear' or 'cyclic'")
+
+    # copy input to avoid changing mutual data
+    signal_1 = signal_1.copy()
+    signal_2 = signal_2.copy()
+
+    # determine signal length for the FFT
+    n_samples = np.array([signal_1.n_samples, signal_2.n_samples], dtype=int)
+    if mode == 'cyclic':
+        if n_samples[0] != n_samples[1]:
+            raise ValueError(("signal_1 and signal_2 must be of the same "
+                              "length in 'cyclic' mode"))
+        n_fft = n_samples[0]
+    else:
+        n_fft = np.sum(n_samples) - 1
+
+        # zero pad signals to match lengths
+        if signal_1.n_samples < n_fft:
+            signal_1 = pyfar.dsp.pad_zeros(
+                signal_1, n_fft - signal_1.n_samples, 'end')
+        if signal_2.n_samples < n_fft:
+            signal_2 = pyfar.dsp.pad_zeros(
+                signal_2, n_fft - signal_2.n_samples, 'beginning')
+
+    # compute correlation as frequency domain convolution
+    # with time flipped and conjugate values for second signal
+    signal_2.time = signal_2.time.copy()[..., ::-1].conj()
+    correlation = signal_1 * signal_2
+    correlation = correlation.time
+
+    # compute lags, i.e., times that the second signal was shifted
+    # with respect to the first
+    if mode == 'cyclic':
+        roll = n_fft // 2
+        correlation = np.roll(correlation, -roll, -1)
+        if n_fft % 2:
+            lags = np.arange(-roll, roll + 1)
+        else:
+            lags = np.arange(-roll + 1, roll + 1)
+    else:
+        lags = np.arange(-n_samples[1] + 1, n_samples[0])
```

Comment:
I'd suggest returning pyfar.TimeData(correlation_data, lags) instead of returning numpy arrays.
I'd also suspect that the lags computation could be replaced by using https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.correlation_lags.html#correlation-lags ?

Reply:
We decided on one of the last meetings that functions should return numpy arrays in such cases, as for example done also in the group delay function. We could discuss this again